dhis-2/dhis-services/dhis-service-tracker/src/main/java/org/hisp/dhis/tracker/validation/hooks/AbstractTrackerValidationHook.java

```diff
@@ -0,0 +1,393 @@
+                .validateAttributeUniquenessWithinScope(
+                    teAttr,
+                    attr.getValue(),
+                    te.getTrackedEntity(),
+                    trackedEntityOu );
+
+            if ( error != null )
+            {
+                errorReporter.addError( newReport( TrackerErrorCode.E1064 ).addArg( error ) );
+                return false;
+            }
+        }
+
+        return true;
+    }
+
+    protected Map<String, TrackedEntityAttributeValue> getTeiAttributeValueMap(
+        List<TrackedEntityAttributeValue> values )
+    {
+        return values.stream().collect( Collectors.toMap( v -> v.getAttribute().getUid(), v -> v ) );
+    }
+
+    protected boolean validateGeo( ValidationErrorReporter errorReporter, TrackedEntity te,
+        FeatureType featureType )
+    {
+        if ( te.getGeometry() != null )
+        {
+            FeatureType typeFromName = FeatureType.getTypeFromName( te.getGeometry().getGeometryType() );
+
+            if ( !featureType.equals( typeFromName ) || FeatureType.NONE == featureType )
+            {
+                errorReporter.addError( newReport( TrackerErrorCode.E1012 )
+                    .withObject( te )
+                    .addArg( featureType.name() ) );
+                return false;
+            }
+        }
+        else
+        {
+            return checkCanCreateGeo( errorReporter, te );
+        }
+
+        return true;
+    }
+
+    protected void validateTrackedEntityType( ValidationErrorReporter errorReporter, TrackerBundle bundle,
+        TrackedEntity te )
+    {
+        if ( bundle.getImportStrategy().isCreate() )
+        {
+            if ( te.getTrackedEntityType() == null )
+            {
+                errorReporter.addError( newReport( TrackerErrorCode.E1004 ).withObject( te ) );
+
+                return;
+            }
+
+            TrackedEntityType entityType = PreheatHelper.getTrackedEntityType( bundle, te.getTrackedEntityType() );
+
+            if ( entityType == null )
+            {
+                errorReporter.addError( newReport( TrackerErrorCode.E1005 )
+                    .withObject( te )
+                    .addArg( te.getTrackedEntityType() ) );
+            }
+        }
+    }
+
+    protected OrganisationUnit getOrganisationUnit( TrackerBundle bundle, TrackedEntity te )
+    {
+        return bundle.getImportStrategy().isCreate() ?
+            PreheatHelper.getOrganisationUnit( bundle, te.getOrgUnit() ) :
+            PreheatHelper.getTrackedEntityInstance( bundle, te.getTrackedEntity() ).getOrganisationUnit();
+    }
+
+    protected TrackedEntityType getTrackedEntityType( TrackerBundle bundle, TrackedEntity te )
+    {
+        return bundle.getImportStrategy().isCreate() ?
+            PreheatHelper.getTrackedEntityType( bundle, te.getTrackedEntityType() ) :
+            PreheatHelper.getTrackedEntityInstance( bundle, te.getTrackedEntity() ).getTrackedEntityType();
+    }
+}
```

Comment:
Overall I would love a bit more comments/javadocs explaining the purpose of each method, and what assumtions you might make in places where its not easily understandable from the code. But code-wise this looks good

Reply:
Yes, agree. I've waited a bit with the documenting until I was sure that the form and structure of the validation hook system was accepted. There has been very much refactoring back and forth. Now that the structure looks like it will stick I will start making proper javadocs to the classes and methods.